File: python/BiosealSdkSampleCLI.py
import base64
import os
from id3bioseal import *

localCache = True
localCacheDir = "cache_bioseal"

def writeFile(path: str, data):
    with open(path, "wb") as f:
        f.write(data)

def writeFileStr(path: str, data : str):
    with open(path, "w") as f:
        f.write(data)

def resourceCallbackHandler(context: Bioseal, args: ResourceCallbackArgs) -> int:
    try:
        if not localCache:
            args.download()
        else:
            print(f"URI = {args.uri}")
            # Cache disque
        if not os.path.exists(localCacheDir):
            os.mkdir(localCacheDir)
        path = os.path.join(localCacheDir, args.resource_name)
        if not os.path.exists(path) or args.requires_update:
            args.download()
            writeFile(path, args.output_data)
        else:
            try:
                with open(path, "rb") as f:
                    data = f.read(-1)
            except IOError:
                return BiosealError.DOWNLOAD_ERROR
            args.output_data = data
        return 0
    except:
        return BiosealError.EXCEPTION_IN_CALLBACK

def displayBioSealInfo(path: str):
    print(f"Decoding BioSeal file {path}")
    try:
        with open(path) as f:
            data = f.read()
            bioseal.decode_from_string(data)
            bioseal.verify()

        print(f"  BioSeal format : {bioseal.format}")
        # display manifest information
        print(f"   Use case: {bioseal.manifest_id:6x}")
        supportedLanguages = bioseal.supported_languages
        for language in supportedLanguages:
            print(f"   Document name: '{bioseal.get_document_name(language)}'")

        print("   Payload:")
        payload = bioseal.payload
        for field_name in payload:
            info = "null"
            field = payload.get(field_name)
            if not field.is_null:
                if field.field_type == FieldType.INTEGER:
                    info = field.value_as_integer
                elif field.field_type == FieldType.STRING:
                    info = field.value_as_string
                elif field.field_type == FieldType.FLOAT:
                    info = field.value_as_float
                elif field.field_type == FieldType.BOOLEAN:
                    info = field.value_as_boolean
                elif field.field_type == FieldType.BINARY:
                    info = base64.b64encode(field.value_as_binary)
                elif field.field_type == FieldType.DATE:
                    info = field.value_as_date.to_string()
                elif field.field_type == FieldType.TIME:
                    info = field.value_as_date.to_string()
                elif field.field_type == FieldType.TIMESTAMP:
                    info = field.value_as_date.to_string()
            print(f"      {field.name}: {field.field_type} = '{info}'")

        # display face template if existing
        hasFaceTemplate = bioseal.contains_face_templates
        print(f"   Face template: {hasFaceTemplate}")
        if hasFaceTemplate:
            fieldFaceList = bioseal.find_biometrics(BiometricDataType.FACIAL_FEATURES, None)
            if fieldFaceList.count > 0:
                fieldFace = fieldFaceList.get(0)
                print("   Face template saved in data folder")
                writeFile(path.replace(".dat", ".template"), fieldFace.value_as_binary)

        # display and save face image if existing
        hasFaceImage = bioseal.contains_portraits
        print(f"   Face image: {hasFaceImage}")
        if hasFaceImage:
            fieldFaceList = bioseal.find_fields_by_extension(FieldExtensionType.PORTRAIT)
            if fieldFaceList.count > 0:
                fieldFace = fieldFaceList.get(0)
                print("   Face image saved in data folder")
                writeFile(path.replace(".dat", ".webp"), fieldFace.value_as_binary)

        # fetch and save presentation view
        print("   Presentation view supported languages:")
        for supportedLanguage in bioseal.supported_languages:
            print(f"      {supportedLanguage}")

        # Get default language
        bioseal.build_html_view(None, True)
        writeFileStr(path.replace(".dat", ".html"), bioseal.html_view)
        print("   Presentation view saved in data folder")

        # build and save JSON file
        jsonPayload = bioseal.build_payload_as_json("  ")
        writeFileStr(path.replace(".dat", ".json"), jsonPayload)
        print("   JSON representation saved in data folder")

        # display signature information
        print( "   Signature:")
        verificationResult = bioseal.verification_result
        print(f"      Signature verified status: {verificationResult.vds_signature_verified}")
        print(f"      Certification chain verified status: {verificationResult.certification_chain_verified}")
        print(f"      Certificate usage authorized status: {verificationResult.signing_certificate_usage_authorized}")

        # display governance information
        print( "   Governance:")
        print(f"      LoTL: {bioseal.lotl_url}")
        print(f"      TSL: {bioseal.tsl_url}")
        print(f"      Manifest: {bioseal.manifest_url}")
        print(f"      LoTL valid status: {verificationResult.lotl_governance_valid}")
        print(f"      TSL valid status: {verificationResult.tsl_governance_valid}")
        print(f"      Manifest valid status: {verificationResult.manifest_governance_valid}")
        print(f"      Authority verified status: {verificationResult.ca_certificate_verified}")

        # display certificate information
        print( "   Certificate:")
        certificateInformation = bioseal.certificate_information
        print(f"      Authority ID: {bioseal.certificate_authority_id}")
        print(f"      Authority issuing country: {bioseal.certificate_authority_issuing_country}")
        print(f"      Issuer: {certificateInformation.issuer_common_name}")
        print(f"      Subject: {certificateInformation.subject_common_name}")
        print(f"      Organization: {certificateInformation.subject_organization}")
        print(f"      Organization unit: {certificateInformation.subject_organizational_unit}")
        print(f"      Date of creation: {certificateInformation.not_before.to_string()}")
        print(f"      Date of expiration: {certificateInformation.not_after.to_string()}")
    except Exception as ex:
        print(ex)

bioseal = Bioseal()
bioseal.external_resource_callback = resourceCallbackHandler

# This basic sample shows how to read BioSeal biographics only contents
displayBioSealInfo("/data/ExBioSealBiographics.dat")

# This sample shows how to read BioSeal face image and template contents
displayBioSealInfo("data/ExBioSealFace.dat")

# This accreditation sample shows how to read BioSeal face template contents
displayBioSealInfo("data/ExBioSealAccreditation.dat")

os.system("pause")

File: java/BiosealSdkSampleCLI.java
import eu.id3.bioseal.*;

import java.io.File;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Base64;

public class BiosealSdkSampleCLI {

    static boolean localCache = true;
    static String localCacheDir = "cache_bioseal";

    static void displayBioSealInfo(Bioseal bioseal, String biosealPath) {
        try {
            System.out.println("Decoding and verification BioSeal file " + biosealPath);
            byte[] bFile = Files.readAllBytes(Paths.get(biosealPath));
            VerificationResult verifyResult =bioseal.verifyFromBuffer(bFile);
            System.out.println("governanceValid = " + verifyResult.governanceValid);
            System.out.println("caCertificateVerified = " + verifyResult.caCertificateVerified);
            System.out.println("certificationChainVerified = " + verifyResult.certificationChainVerified);
            System.out.println("signingCertificateUsageAuthorized = " + verifyResult.signingCertificateUsageAuthorized);
            System.out.println("vdsNotExpired = " + verifyResult.vdsNotExpired);
            System.out.println("vdsSignatureVerified = " + verifyResult.vdsSignatureVerified);

            System.out.println("  BioSeal format : " + bioseal.getFormat());
            // display manifest information
            System.out.println("   Use case: " + String.format("%06x", bioseal.getManifestId()));
            var supportedLanguages = bioseal.getSupportedLanguages();
            for (var language : supportedLanguages) {
                System.out.println("   Document name: '"+ bioseal.getDocumentName(language)+ "'");
            }

            System.out.println("   Payload:");
            var payload = bioseal.getPayload();
            for (var field : payload) {
                String info = "null";
                if (!field.getIsNull() ) {
                    switch (field.getFieldType()) {
                        case INTEGER -> info = String.valueOf(field.getValueAsInteger());
                        case STRING -> info = field.getValueAsString();
                        case FLOAT -> info = String.valueOf(field.getValueAsFloat());
                        case BOOLEAN -> info = String.valueOf(field.getValueAsBoolean());
                        case BINARY -> info = Base64.getEncoder().encodeToString(field.getValueAsBinary());
                        case DATE -> info = field.getValueAsDate().toString();
                        case TIME -> info = field.getValueAsTime().toString();
                        case TIMESTAMP -> info = field.getValueAsDateTime().toString();
                        default -> info = "?";
                    }
                }
                System.out.println("      " + field.getName() + ": " + field.getFieldType() + " = '" + info + "'");
            }

            // display face template if existing
            boolean hasFaceTemplate = bioseal.getContainsFaceTemplates();
            System.out.println("   Face template: " + hasFaceTemplate);
            if (hasFaceTemplate) {
                var fieldFaceList = bioseal.findBiometrics(BiometricDataType.FACIAL_FEATURES, null);
                if (fieldFaceList.getCount() > 0) {
                    var fieldFace = fieldFaceList.get(0);
                    System.out.println("   Face template saved in data folder");
                    Files.write(Paths.get(biosealPath.replace(".dat", ".template")), fieldFace.getValueAsBinary());
                }
            }

            // display and save face image if existing
            boolean hasFaceImage = bioseal.getContainsPortraits();
            System.out.println("   Face image: " + hasFaceImage);
            if (hasFaceImage) {
                var fieldFaceList = bioseal.findFieldsByExtension(FieldExtensionType.PORTRAIT);
                if (fieldFaceList.getCount() > 0) {
                    var fieldFace = fieldFaceList.get(0);
                    System.out.println("   Face image saved in data folder");
                    Files.write(Paths.get(biosealPath.replace(".dat", ".webp")), fieldFace.getValueAsBinary());
                }
            }

            // fetch and save presentation view
            System.out.println("   Presentation view supported languages:");
            supportedLanguages = bioseal.getSupportedLanguages();
            for (var supportedLanguage : supportedLanguages) {
                System.out.println("      " + supportedLanguage);
            }

            // Get default language
            bioseal.buildHtmlView(null, true);
            Files.writeString(Paths.get(biosealPath.replace(".dat", ".html")), bioseal.getHtmlView());
            System.out.println("   Presentation view saved in data folder");

            // build and save JSON file
            var jsonPayload = bioseal.buildVdsAsJson("  ");
            Files.writeString(Paths.get(biosealPath.replace(".dat", ".json")), jsonPayload);
            System.out.println("   JSON representation saved in data folder");

            // display signature information
            var verificationResult = bioseal.getVerificationResult();
            System.out.println("   Signature:");
            System.out.println("      Signature verified status: " + verificationResult.vdsSignatureVerified);
            System.out.println("      Certification chain verified status: " + verificationResult.certificationChainVerified);
            System.out.println("      Certificate usage authorized status: " + verificationResult.signingCertificateUsageAuthorized);

            // display governance information
            System.out.println("   Governance:");
            System.out.println("      LoTL: " + bioseal.getLotlUrl());
            System.out.println("      TSL: " + bioseal.getTslUrl());
            System.out.println("      Manifest: " + bioseal.getManifestUrl());
            System.out.println("      LoTL valid status: " + verificationResult.lotlGovernanceValid);
            System.out.println("      TSL valid status: " + verificationResult.tslGovernanceValid);
            System.out.println("      Manifest valid status: " + verificationResult.manifestGovernanceValid);
            System.out.println("      Authority verified status: " + verificationResult.caCertificateVerified);

            // display certificate information
            var certificateInformation = bioseal.getCertificateInformation();
            System.out.println("   Certificate:");
            System.out.println("      Authority AC: " + bioseal.getCertificateAuthorityReference());
            System.out.println("      Authority ID: " + bioseal.getCertificateIdentifier());
            System.out.println("      Issuer: " + certificateInformation.getIssuerCommonName());
            System.out.println("      Subject: " + certificateInformation.getSubjectCommonName());
            System.out.println("      Organization: " + certificateInformation.getSubjectOrganization());
            System.out.println("      Organization unit: " + certificateInformation.getSubjectOrganizationalUnit());
            System.out.println("      Date of creation: " + certificateInformation.getNotBefore().toString());
            System.out.println("      Date of expiration: " + certificateInformation.getNotAfter().toString());

        }
        catch (Exception ex) {
            System.out.println(ex.getMessage());
        }
    }

    public static void main(String[] args) {
        System.out.println("---------------------------------------");
        System.out.println("id3.Samples.Bioseal.BiosealSdkSampleCLI");
        System.out.println("---------------------------------------");

        System.out.println(BiosealLibrary.getVersion());

        try {
            try (Bioseal bioseal = new Bioseal()) {

                ResourceCallbackHandler callback1 = (bioseal1, args1) -> {
                    try {
                        if (!localCache) {
                            args1.download();
                        }
                        else {
                            File dir = new File(localCacheDir);
                            if(!dir.exists())  {
                                dir.mkdir();
                            }
                            File cacheFile = new File(localCacheDir + File.separator + args1.getResourceName());
                            if (!cacheFile.exists() || args1.getRequiresUpdate()) {
                                args1.download();
                                Files.write(Paths.get(localCacheDir + File.separator + args1.getResourceName()), args1.getOutputData());
                            }
                            else {
                                args1.setOutputData(Files.readAllBytes(Paths.get(localCacheDir + File.separator + args1.getResourceName())));
                            }
                        }
                    }
                    catch (Exception ex) {
                        return BiosealError.EXCEPTION_IN_CALLBACK.getValue();
                    }
                    return 0;
                };
                bioseal.setExternalResourceCallback(callback1);
                displayBioSealInfo(bioseal, "../data/ExBioSealBiographics.dat");
                displayBioSealInfo(bioseal, "../data/ExBioSealFace.dat");
                displayBioSealInfo(bioseal, "../data/ExBioSealAccreditation.dat");
            }
            catch (BiosealException ex) {
                System.out.println(ex.getMessage());
            }
        }
        catch (Exception ex) {
            System.out.println(ex.getMessage());
        }

        System.out.println("Sample terminated successfully.");
    }
}
